extension/jmxmetricsextension/subprocess/subprocess.go

```diff
@@ -0,0 +1,310 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package subprocess
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"sync"
+	"syscall"
+	"time"
+
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+// exported to be used by jmx metrics extension
+type Config struct {
+	ExecutablePath       string            `mapstructure:"executable_path"`
+	Args                 []string          `mapstructure:"args"`
+	EnvironmentVariables map[string]string `mapstructure:"environment_variables"`
+	StdInContents        string            `mapstructure:"stdin_contents"`
+	RestartOnError       bool              `mapstructure:"restart_on_error"`
+	RestartDelay         *time.Duration    `mapstructure:"restart_delay"`
+	ShutdownTimeout      *time.Duration    `mapstructure:"shutdown_timeout"`
+}
+
+// exported to be used by jmx metrics extension
+type Subprocess struct {
+	Stdout         chan string
+	cancel         context.CancelFunc
+	config         *Config
+	envVars        []string
+	logger         *zap.Logger
+	pid            *pid
```

Comment:
Does this need to be a pointer to `pid` or can be a non-pointer field?

Reply:
Since there's just one associated w/ the process it can use the value but the zero value of its represented process id will be incorrect (no user process can have a pid of 0).  Updated to return -1 if nil if that's ok with you.

Reply:
I thought about this and the same result should probably be done w/ a 0 check. updated.